### ISSUE TITLE
nodelocaldns forever crashing/restarting [Info/Solution]

### ISSUE THREAD

**Issue by Steven9Smith** [NVIDIA/deepops#1285]
This is more of a kubernetes configuration issue with the kubelet. here is the chat i had with Chat GPT to fix this. the solution will be at the end:
ChatGPT
The /etc/resolv.conf configurations on both nodes seem fine now. They're set to use Google's DNS servers (8.8.8.8 and 8.8.4.4) instead of the local resolver (127.0.0.53). This should help avoid the DNS loop issue.

However, keep in mind that in a Kubernetes environment, the DNS configurations within the nodes don't affect the DNS resolution inside the pods directly. The DNS for pods is handled by CoreDNS or any other DNS service running in your Kubernetes cluster, and it's configured separately.

Given the change in the nodes' /etc/resolv.conf, you should now check the state of your nodelocaldns pod. If it's still not running correctly, you might want to look at the ConfigMap for the nodelocaldns again, and make sure it's set up correctly.

You can get the details of the nodelocaldns pod using:

```bash

kubectl -n kube-system get pods -l k8s-app=node-local-dns -o wide
```
And to see the logs of a specific nodelocaldns pod:

```bash

kubectl -n kube-system logs <pod-name>
```
Replace <pod-name> with the name of your nodelocaldns pod.

If there's still an issue related to Corefile.base, you might need to adjust the configuration of your nodelocaldns to either provide this file or change the nodelocaldns to not expect this file. It would be best to refer to the specific nodelocaldns documentation or the configuration guide provided with the version of nodelocaldns you are using.
User
how do i perform this instruction? "Add the following to kubelet: --resolv-conf <path-to-your-real-resolv-conf-file>. Your "real" resolv.conf is the one that contains the actual IPs of your upstream servers, and no local/loopback address. This flag tells kubelet to pass an alternate resolv.conf to Pods. For systems using systemd-resolved, /run/systemd/resolve/resolv.conf is typically the location of the "real" resolv.conf, although this can be different depending on your distribution.". i need the commands and steps
ChatGPT

To add the --resolv-conf flag to the kubelet, you will need to modify the kubelet configuration file. The exact location of this file may vary depending on your Kubernetes setup, but it's typically located at /etc/systemd/system/kubelet.service.d/10-kubeadm.conf.

Here are the general steps to add the --resolv-conf flag to the kubelet:

    Open the kubelet configuration file in a text editor. If your file is located at /etc/systemd/system/kubelet.service.d/10-kubeadm.conf, you would use the following command:

```bash

sudo vim /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
```
    Look for a line that starts with ExecStart=/usr/bin/kubelet. This line contains the command that is used to start the kubelet and any flags that are passed to it.

    Add the --resolv-conf flag to the end of the ExecStart line, followed by the path to your actual resolv.conf file. It should look something like this:

```bash

ExecStart=/usr/bin/kubelet ... --resolv-conf=/run/systemd/resolve/resolv.conf
```
    Save the file and exit the text editor. If you're using vim, you can do this by pressing Esc and then typing :wq followed by Enter.

    Restart the kubelet to apply the changes:

```bash

sudo systemctl daemon-reload
sudo systemctl restart kubelet
```
    Verify that the new flag has been applied by checking the kubelet's active configuration:

```bash

ps aux | grep kubelet
```
In the output, you should see the --resolv-conf flag with the correct path.

Please note that these instructions assume your Kubernetes setup uses systemd and kubelet is running as a systemd service. The exact steps may vary depending on your specific Kubernetes setup. Always make sure to have a backup of any configuration file before modifying it.

NOTE: if your don't want to edit using vim use 
`export EDITOR=nano` 

also make sure your /etc/resolv.conf looks something like

```
nameserver 8.8.8.8
nameserver 8.8.4.4 
```

when your done restart coredns and localdnsnode on the pods in kube-system that were having issue (or the pods with issues in another namespace name)

**Comment by cocakohler** [comment]
@Steven9Smith i had the same issue (Pod crashed with OOM). I adapted the request/limits of the nodelocaldns, since then it is runnig fine again

**Comment by dholt** [comment]
Following up on this stale-closed issue as part of the resumed DeepOps release cadence — and thanks for posting your solution in-thread, it was exactly right: the crashloop is CoreDNS's `loop` plugin correctly detecting a DNS forward loop when the host resolv.conf points at the systemd-resolved stub (127.0.0.53).

Good news: your kubelet `--resolv-conf` fix is now the automatic default. Current DeepOps (Kubespray v2.31) sets `kube_resolv_conf: /run/systemd/resolve/resolv.conf` on all systemd-resolved OSes (Ubuntu 18/20/22/24, Fedora), so kubelet reads the real upstream resolvers instead of the loopback stub and nodelocaldns starts cleanly. For the separate OOM variant reported later in this thread, the resource knobs are exposed as `nodelocaldns_memory_limit` / `nodelocaldns_memory_requests` overrides. Present in releases 26.05+. If you see a nodelocaldns crashloop on a current release, please open a new issue.
